### PR TITLE
PSR-7 Part 2: Optimize HTTPRequest

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #10619 and #10631 

Changes:
- Rename `(I)HTTPRequest` into `(I)HTTPClient`
- Move Guzzle-`Client` initialization into a Factory with all default parameters
-  `HTTPClient::get()`, `HTTPClient::head()`, `HTTPClient::post()` are now using the same codebase per `HTTPClient::request()`
- Use `mattwright/urlresolver` for easy `finalUrl` resolving with some more redirect possibilities
- Created `DiceHttpMockHandlerTrait` for easy mocking requests and responses of HTTP endpoints
- Created a `ImagesTest` to mock remote image fetching per `getInfoFromURL()` (this would have shown the bug of last year)
- Created, but currently skipped `ProbeTest::testCleanUri()` because of too much requests (this would have shown the bug of #10631 )
- Fixes `GServer` fetching:  https://github.com/nupplaphil/friendica/commit/e8054703a68be5ed0738179ca8556585da177525 
- Fixes https://github.com/nupplaphil/friendica/commit/e8054703a68be5ed0738179ca8556585da177525
- Fixes usage of headers (map instead of array!)

This should be the last step before we can now easily extend the `HTTPClient` for `PUT` and `PROP` requests for the WebDav Backend :-)

ToDos:
- [x] CleanUp/Rebase commits (reword, squash)
- [x] As usual --> running on my node with logging as much as possible

### CODE REVIEW

File: src/Factory/HTTPClientFactory.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			RequestOptions::HTTP_ERRORS => false,
+			// Without this setting it seems as if some webservers send compressed content
+			// This seems to confuse curl so that it shows this uncompressed.
+			/// @todo  We could possibly set this value to "gzip" or something similar
+			RequestOptions::DECODE_CONTENT   => '',
+			RequestOptions::FORCE_IP_RESOLVE => ($this->config->get('system', 'ipv4_resolve') ? 'v4' : null),
+			RequestOptions::CONNECT_TIMEOUT  => 10,
+			RequestOptions::TIMEOUT          => $this->config->get('system', 'curl_timeout', 60),
+			// by default we will allow self-signed certs
+			// but you can override this
```
Comment (MrPetovan) [inline src/Factory/HTTPClientFactory.php]:
```suggestion
			// but it can be overridden
```

File: src/Network/HTTPClient.php [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+				$response->getHeaderLine('Content-Length') > $opts['content_length']) {
+				throw new TransferException('The file is too big!');
+			}
+		};
+
+		try {
+			switch ($method) {
+				case 'get':
+				case 'head':
+				case 'post':
```
Comment (MrPetovan) [inline src/Network/HTTPClient.php:137]: PUT and DELETE are pretty popular too.
```suggestion
				case 'get':
				case 'head':
				case 'post':
				case 'put':
				case 'delete':
```

File: tests/src/Network/ProbeTest.php [hunk 3]
```diff
@@ -105,4 +116,122 @@ public function testGetFeedLinkBase()
...
+			]
+		];
+	}
+
+	/**
+	 * @dataProvider dataUri
+	 */
+	public function testCleanUri(string $uri, string $assertUri, array $assertInfos)
+	{
+		self::markTestIncomplete('hard work due mocking 19 different http-requests');
```
Comment (MrPetovan) [inline tests/src/Network/ProbeTest.php:168]: `Network\Probe::cleanUri()` doesn't perform any remote HTTP calls, are you confusing with `Network\Probe::uri()`?

**Review by MrPetovan — APPROVED** [review]

File: tests/src/Network/ProbeTest.php (hunk 3 above)
Comment (nupplaphil) [reply tests/src/Network/ProbeTest.php:168]: yes indeed, I merged both tests :-)

File: src/Network/IHTTPClient.php [hunk 4]
```diff
@@ -61,7 +61,7 @@ public function fetchFull(string $url, int $timeout = 0, string $accept_content
 	 *
 	 * @param string $url        URL to fetch
 	 * @param array  $opts       (optional parameters) assoziative array with:
```
Comment (MrPetovan) [inline src/Network/IHTTPClient.php]:
```suggestion
	 * @param array  $opts       (optional parameters) associative array with:
```

File: src/Network/IHTTPClient.php [hunk 5]
```diff
@@ -75,7 +75,7 @@ public function head(string $url, array $opts = []);
...
```
Comment (MrPetovan) [inline src/Network/IHTTPClient.php]:
```suggestion
	 * @param array  $opts       (optional parameters) associative array with:
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]
Pro Tip: In the "Files" tab you can combine several suggestions in a single commit.

**Comment by nupplaphil** [comment]
The only failures so far were some re-usages of the cookie for the url-resolver, I added the `overwrite` parameter, so this case should be covered now

--> ready for merge

**Comment by MrPetovan** [comment]
I can't see the `overwrite` parameter anywhere in the diff?

**Comment by nupplaphil** [comment]
https://github.com/friendica/friendica/pull/10633/commits/4ddaf49f537b6e79aac3f11f98440329aa3221c5#diff-6a8e207b35f84340e9a4fff57afe5d5c385218b21b93ee42a2b56c3991aee132R106
second paramter `true` :-)

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**